thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobScheduler.java

```diff
@@ -1,274 +1,268 @@
-
-    try {
-      quartzScheduler.scheduleJob(job, trigger);
-    } catch (SchedulerException e) {
-      LOG.error("Exception while scheduling job", e);
+    boolean pass = checkIfDetectionRunCriteriaMet(startTime, endTime, datasetConfig, anomalyFunction);
+    if (pass) {
+      jobExecutionId = runAnomalyFunctionOnRanges(anomalyFunction, Lists.newArrayList(startTime), Lists.newArrayList(endTime));
+    } else {
+      LOG.warn("Data incomplete for monitoring window {} ({}) to {} ({}), skipping anomaly detection",
+          startTime, new DateTime(startTime), endTime, new DateTime(endTime));
+      // TODO: Send email to owners/dev team
     }
-
-    LOG.info("Started {}: {}", jobKey, anomalyFunctionSpec);
+    return jobExecutionId;
   }
 
-  private List<AnomalyFunctionDTO> readAnomalyFunctionSpecs() {
-    return DAO_REGISTRY.getAnomalyFunctionDAO().findAll();
+  /**
+   * Checks if a time range for a dataset meets data completeness criteria
+   * @param startTime
+   * @param endTime
+   * @param datasetConfig
+   * @param anomalyFunction
+   * @return
+   */
+  private boolean checkIfDetectionRunCriteriaMet(Long startTime, Long endTime, DatasetConfigDTO datasetConfig, AnomalyFunctionDTO anomalyFunction) {
+    boolean pass = true;
+    String dataset = datasetConfig.getDataset();
+
+    /**
+     * Check is completeness check required is set at dataset level. That flag is false by default, so user will set as needed
+     * Check also for same flag in function level. That flag is true by default, so dataset config's flag will have its way unless user has tampered with this flag
+     * This flag would typically be unset, in backfill cases
+     */
+    if (datasetConfig.isRequiresCompletenessCheck() && anomalyFunction.isRequiresCompletenessCheck()) {
+
+      LOG.info("Checking for completeness for dataset {} and time range {}({}) to {}({})", dataset,
+          startTime, new DateTime(startTime), endTime, new DateTime(endTime));
+
+      List<DataCompletenessConfigDTO> incompleteTimePeriods = DAO_REGISTRY.getDataCompletenessConfigDAO().
+          findAllByDatasetAndInTimeRangeAndStatus(dataset, startTime, endTime, false);
+      LOG.info("Incomplete periods {}", incompleteTimePeriods);
+      List<DataCompletenessConfigDTO> completeTimePeriods = DAO_REGISTRY.getDataCompletenessConfigDAO().
+          findAllByDatasetAndInTimeRangeAndStatus(dataset, startTime, endTime, true);
+      LOG.info("Complete periods {}", completeTimePeriods);
+      long expectedCompleteBuckets = DetectionJobSchedulerUtils.getExpectedCompleteBuckets(datasetConfig, startTime, endTime);
+      LOG.info("Num complete periods: {} Expected num buckets:{}", completeTimePeriods.size(), expectedCompleteBuckets);
+
+      // if available, run the function
+      if (incompleteTimePeriods.size() == 0 && // nothing incomplete
+          completeTimePeriods.size() == expectedCompleteBuckets) { // complete matches expected buckets
```

Comment:
Two suggestions here:
1. (Minor) Move lines 223-227 inside the condition at line 230 to optimize the number we access DB. Hence, we only access DB for completeTimePeriod only if imcompleteTimePeriods is 0. 
2. Add comment to document the goal of this if-statement. Just in case we forgot what is happening here after some time. An example the show how the if-statement is used should suffice.